# CONVOLUTION NEURAL NETWORKS (CNN) PRACTICE

In [ ]:
# Loading Libraries
from tensorflow.keras.datasets import mnist #Dataset
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense # layer for CNN
from tensorflow.keras.models import Sequential #stack of layers
from tensorflow.keras.utils import to_categorical # converts labels into one-hot encoded

In [ ]:
# Loading the dataset
(X_train, y_train), (X_test, y_test) = mnist.load_data()

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


In [ ]:
# Preprocessing the dataset
X_train = X_train.reshape(-1, 28, 28, 1).astype('float32')/255
X_test = X_test.reshape(-1, 28, 28, 1).astype('float32')/255


In [ ]:
# Converts labels into one hot encoding
y_train = to_categorical (y_train)
y_test = to_categorical (y_test)

In [ ]:
# Constructing the CNN
model = Sequential([
    Conv2D(32, (3, 3), activation= 'relu', input_shape= (28, 28, 1)),
    MaxPooling2D(pool_size=(2, 2)),

    Conv2D(64, (3, 3), activation= 'relu', input_shape= (28, 28, 1)),
    MaxPooling2D(pool_size=(2, 2)),

    Flatten(),
    Dense(128, activation= 'relu'),
    Dense(10, activation= 'softmax')
])

/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
# Compiling and training the model
model.compile(optimizer='Adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, epochs=5, batch_size=32, validation_split=0.1)

Epoch 1/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 12s 4ms/step - accuracy: 0.9048 - loss: 0.3118 - val_accuracy: 0.9855 - val_loss: 0.0523
Epoch 2/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.9863 - loss: 0.0414 - val_accuracy: 0.9900 - val_loss: 0.0358
Epoch 3/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9915 - loss: 0.0272 - val_accuracy: 0.9895 - val_loss: 0.0395
Epoch 4/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9934 - loss: 0.0196 - val_accuracy: 0.9902 - val_loss: 0.0369
Epoch 5/5
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9953 - loss: 0.0146 - val_accuracy: 0.9905 - val_loss: 0.0377


In [ ]:
# Evaluating Test Data
loss, acc = model.evaluate(X_test, y_test)
print(f'Test Loss: {loss}')
print(f'Test Accuracy: {acc}')

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9848 - loss: 0.0448
Test Loss: 0.035831984132528305
Test Accuracy: 0.9883999824523926


SUMMARY:

accuracy: 0.9848 → While training, the final accuracy on the last epoch’s validation (or evaluation) was 98.48%.

loss: 0.0448 → The training (or validation) loss at the end of the last epoch was 0.0448. Lower is better.

Test Loss: 0.0358 → When we ran the model on the held-out test set, the model’s loss was 0.0358, which is even better than the training loss (good sign—no overfitting).

Test Accuracy: 0.9884 → The model correctly classified about 98.84% of the test digits.

In simple terms: the CNN did an excellent job — nearly 99% accuracy on unseen MNIST data.